# Bioproject PRJEB14440

This sample comes from the [NCBI BioProject PRJEB14440](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJEB14440).

# Title:
Pseusomonas aeruginosa metagenomes

_Is genotyping of single isolates sufficient for population structure analysis of Pseudomonas aeruginosa in cystic fibrosis airways?_

# Description:
BackgroundThe primary cause of morbidity and mortality in cystic fibrosis (CF) patients is lung infection by Pseudomonas aeruginosa. Therefore much work has been done to understand the adaptation and evolution of P. aeruginosa in the CF lung. However, many of these studies have focused on longitudinally collected single isolates, and only few have included cross-sectional analyses of entire P. aeruginosa populations in sputum samples. To date only few studies have used the approach of metagenomic analysis for the purpose of investigating P. aeruginosa populations in CF airways. ResultsWe analysed five metagenomes together with longitudinally collected single isolates from four recently chronically infected CF patients. With this approach we were able to link the clone type and the majority of SNP profiles of the single isolates to that of the metagenome(s) for each individual patient. ConclusionBased on our analysis we find that when having access to comprehensive collections of longitudinal single isolates it is possible to rediscover the genotypes of the single isolates in the metagenomic samples. This suggests that information gained from genome sequencing of comprehensive collections of single isolates is satisfactory for many investigations of adaptation and evolution of P. aeruginosa to the CF airways.


# Metadata:
We have included the metadata in two separate files, a [JSON format file](PRJEB14440.metadata.json.gz) that is computer readable, or a [tsv format file](PRJEB14440.metadata.tsv.gz) that you can import into excel or similar.


# Publications

Sommer LM, Marvig RL, Luja¡n A, Koza A, Pressler T, Molin S, Johansen HK. 2016. Is genotyping of single isolates sufficient for population structure analysis of Pseudomonas aeruginosa in cystic fibrosis airways? [BMC Genomics 17:589](https://pmc.ncbi.nlm.nih.gov/articles/PMC4979127/).

# Analysis:

- The entire data set is 5 runs, and 1,330,576,074 bp (including amplicon sequences)
- We analysed 5 metagenomic sequence runs.
- We predicted 5 samples out of 5 (100.0%) have _Pseudomonas aeruginosa_


## Prediction outcomes

Sample | Pseudomonas Prediction | Confidence | Certainty
 --- | --- | --- | ---
ERR1457158 | Positive | 0.72 | Medium
ERR1457159 | Positive | 0.72 | Medium
ERR1457160 | Positive | 0.68 | Medium
ERR1457161 | Positive | 0.75 | Medium
ERR1457162 | Positive | 0.72 | Medium


## t-SNE
![Comparison of Adelaide and PRJEB14440 samples by t-SNE](img/PRJEB14440_Pseudomonas_tSNE.png 'Fig. t-SNE of all the analysed sequence data coloured by whether Pseudomonas is predicted')


## PCA
![This cluster of features are most strongly associated with the presence of Pseudomonas](img/PRJEB14440_Pseudomonas_PCA.png 'Fig. PCA of the cluster of features most strongly associated with Pseudomonas colonization in PRJEB14440')


