{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "61210b66-40df-4a56-bcd4-5fc57b11f686",
   "metadata": {},
   "source": [
    "# Worldwide data\n",
    "\n",
    "Read the worldwide data, and then see if we can predict anything about them!\n",
    "\n",
    "Note, cells 3 & 4 get a list of all the samples.\n",
    "\n",
    "These lines in cell 4 are key:\n",
    "\n",
    "```\n",
    "# so far, PRJEB20836 has the best\n",
    "sample_id = worldwide_samples[3]\n",
    "print(f\"Sample is {sample_id}\", file=sys.stderr)\n",
    "wwdf, wwmd = cf_analysis_lib.read_worldwide_data(sample_id)\n",
    "wwdf\n",
    "```\n",
    "\n",
    "You just change the index of the `worldwide_sample_number` to reflect the one you are interested in, and restart the kernel and run all cells.\n",
    "\n",
    "**_NOTE:_** This notebook is explicitly for PRJNA510441 because it has the SHAP waterfall plots at the end!\n",
    "\n",
    "**_NOTE2:_** This notebook uses the MinION model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b46f2774-f4da-4bb3-b72d-86eab08e99d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "worldwide_sample_number = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a5d55948-9d38-44d7-bc77-35661838c080",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "from socket import gethostname\n",
    "\n",
    "hostname = gethostname()\n",
    "\n",
    "if hostname.startswith('hpc-node'):\n",
    "    IN_DEEPTHOUGHT = True\n",
    "    sys.path.append('..')\n",
    "else:\n",
    "    IN_DEEPTHOUGHT = False\n",
    "from cf_analysis_lib.load_libraries import *\n",
    "import cf_analysis_lib\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c239aee6-8526-44b6-97ec-e52baa19cfa7",
   "metadata": {},
   "source": [
    "# Read our data and models\n",
    "\n",
    "This is our normal models we use for predicting everything"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d9130024-e29f-4e5a-8a42-b8e544b902b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Using ../MinION/FunctionalAnalysis/subsystems/subsystems_norm_ss.tsv.gz for the subsystems\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(61, 150)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sequence_type = \"MinION\"\n",
    "datadir = '..'\n",
    "\n",
    "sslevel='subsystems'\n",
    "ss_normalisation='norm_ss'\n",
    "taxa = \"family\"\n",
    "# all_taxa: True = everything, False = bacteria\n",
    "all_taxa = False \n",
    "\n",
    "our_df, our_metadata = cf_analysis_lib.read_the_data(sequence_type, datadir, all_taxa=all_taxa, sslevel=f\"{sslevel}_{ss_normalisation}.tsv.gz\", taxa=taxa)\n",
    "\n",
    "#encoder_models = 'cluster_gbrfs_eukaryotes'\n",
    "encoder_models = 'cluster_gbrfs_minion'\n",
    "if not os.path.exists(os.path.join(encoder_models, 'clusters.json')):\n",
    "    print(\"Please run the autoencoder code before trying to load the models.\", file=sys.stderr)\n",
    "    exit(1)\n",
    "\n",
    "if not os.path.exists(os.path.join(encoder_models, 'pc_df.tsv')):\n",
    "    print(\"Please create and save the PCA before trying to load the models.\", file=sys.stderr)\n",
    "    exit(1)\n",
    "\n",
    "with open(os.path.join(encoder_models, 'clusters.json'), 'r') as file:\n",
    "    data = json.load(file)\n",
    "tmpjsondf = pd.DataFrame(list(data.items()), columns=['Cluster', 'Feature'])\n",
    "cluster_assignments = tmpjsondf.explode('Feature').reset_index(drop=True)\n",
    "cluster_assignments['Cluster'] = cluster_assignments['Cluster'].astype(int)\n",
    "\n",
    "pc_df = pd.read_csv(os.path.join(encoder_models, 'pc_df.tsv'), sep=\"\\t\", index_col=0)\n",
    "pc_df.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62bf90bb-28e1-4cd1-83a6-c511e5262534",
   "metadata": {},
   "source": [
    "# Read one of the data sets\n",
    "\n",
    "Samples: \n",
    "['PRJNA1081394', 'PRJNA516442', 'PRJEB51171', 'PRJNA71831', 'PRJEB20836', 'PRJNA516870', 'PRJNA1126024', 'PRJEB14440', 'PRJEB32062', 'PRJEB54014', 'PRJNA1055940', 'PRJNA1091195', 'PRJNA1101448', 'PRJNA316056', 'PRJNA316588', 'PRJNA510441', 'PRJNA615628', 'PRJNA644285', 'PRJNA825831', 'PRJNA839435', 'PRJNA846291', 'PRJNA931830']\n",
    "\n",
    "Initially, lets see what's there?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1f46e9c0-9a34-454b-b370-9807b649b145",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 24\n",
      "['PRJNA1081394', 'PRJNA516442', 'PRJEB51171', 'PRJNA71831', 'PRJEB20836', 'PRJNA516870', 'PRJNA1126024', 'PRJEB14440', 'PRJEB32062', 'PRJEB54014', 'PRJNA1055940', 'PRJNA1091195', 'PRJNA1101448', 'PRJNA316056', 'PRJNA316588', 'PRJNA510441', 'PRJNA615628', 'PRJNA644285', 'PRJNA825831', 'PRJNA839435', 'PRJNA846291', 'PRJNA931830', 'not_analysed', 'papers']\n"
     ]
    }
   ],
   "source": [
    "worldwide_samples = cf_analysis_lib.worldwide_samples()\n",
    "print(f\"There are {len(worldwide_samples)}\")\n",
    "print(worldwide_samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1a2715d1-854a-48a4-82a1-d490c10faaef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find one sample\n",
    "worldwide_samples.index('PRJNA510441')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e589b8f-633d-41ac-9a9e-1985ddf6f30c",
   "metadata": {},
   "source": [
    "### Read the ww data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "784b4c26-7492-4a23-ab03-52be566ce19d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"background-color:#d4edda;border:1px solid #155724;color:#155724;padding:10px 12px;border-radius:8px;font-family:system-ui,-apple-system,Segoe UI,Roboto,Helvetica,Arial,sans-serif;line-height:1.4;max-width:none;\"><div style=\"font-weight:600; margin-bottom:4px;\">✅ Analysis Report</div><div>Analysing: PRJNA510441</div></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Sample is PRJNA510441\n",
      "Reading subsystems from ../WorldWideDataAnalysis/Atavide/PRJNA510441/subsystems/PRJNA510441_subsystems_norm_ss.tsv.gz\n",
      "Before dropping suspects, shape is (615, 14) and after is (615, 14)\n",
      "Read 14 samples and 615 subsystems\n",
      "Reading taxonomy from ../WorldWideDataAnalysis/Atavide/PRJNA510441/taxonomy_summary/PRJNA510441_family.norm.tsv.gz\n",
      "Read 14 samples and 0 family\n",
      "Read 14 samples and 7 metadata columns\n",
      "We added 433 columns to ww df that were missing\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2,3-diacetamido-2,3-dideoxy-d-mannuronic acid</th>\n",
       "      <th>2-O-alpha-mannosyl-D-glycerate utilization</th>\n",
       "      <th>2-aminophenol Metabolism</th>\n",
       "      <th>2-ketoacid oxidoreductases disambiguation</th>\n",
       "      <th>2-oxoglutarate dehydrogenase</th>\n",
       "      <th>2-phosphoglycolate salvage</th>\n",
       "      <th>3-amino-5-hydroxybenzoic Acid Synthesis</th>\n",
       "      <th>5-methylaminomethyl-2-thiouridine</th>\n",
       "      <th>A Hypothetical Protein Related to Proline Metabolism</th>\n",
       "      <th>A new toxin - antitoxin system</th>\n",
       "      <th>...</th>\n",
       "      <th>Thermodesulfobiaceae</th>\n",
       "      <th>Petrotogaceae</th>\n",
       "      <th>Thermotogaceae</th>\n",
       "      <th></th>\n",
       "      <th>Methylacidiphilaceae</th>\n",
       "      <th></th>\n",
       "      <th>Akkermansiaceae</th>\n",
       "      <th>Verrucomicrobiaceae</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR8334087</th>\n",
       "      <td>49.467681</td>\n",
       "      <td>1879.771864</td>\n",
       "      <td>445.209126</td>\n",
       "      <td>676.058302</td>\n",
       "      <td>123.669202</td>\n",
       "      <td>321.539924</td>\n",
       "      <td>354.518378</td>\n",
       "      <td>4996.235745</td>\n",
       "      <td>700.792142</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334088</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1556.179701</td>\n",
       "      <td>552.192797</td>\n",
       "      <td>363.945253</td>\n",
       "      <td>192.430823</td>\n",
       "      <td>363.945253</td>\n",
       "      <td>332.570662</td>\n",
       "      <td>3714.751544</td>\n",
       "      <td>811.556080</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334089</th>\n",
       "      <td>77.604605</td>\n",
       "      <td>1034.728061</td>\n",
       "      <td>724.309642</td>\n",
       "      <td>120.718274</td>\n",
       "      <td>564.789066</td>\n",
       "      <td>659.639139</td>\n",
       "      <td>482.873095</td>\n",
       "      <td>2819.633965</td>\n",
       "      <td>379.400289</td>\n",
       "      <td>51.736403</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334090</th>\n",
       "      <td>27.080198</td>\n",
       "      <td>1299.849486</td>\n",
       "      <td>649.924743</td>\n",
       "      <td>45.133663</td>\n",
       "      <td>344.821183</td>\n",
       "      <td>352.042569</td>\n",
       "      <td>333.989104</td>\n",
       "      <td>3709.987074</td>\n",
       "      <td>704.085138</td>\n",
       "      <td>162.481186</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334091</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>5327.887049</td>\n",
       "      <td>5327.887049</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1331.971762</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334092</th>\n",
       "      <td>21.559847</td>\n",
       "      <td>2845.899764</td>\n",
       "      <td>344.957547</td>\n",
       "      <td>172.478774</td>\n",
       "      <td>141.576327</td>\n",
       "      <td>711.474941</td>\n",
       "      <td>463.536704</td>\n",
       "      <td>4311.969339</td>\n",
       "      <td>596.489092</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334093</th>\n",
       "      <td>364.811340</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>625.390869</td>\n",
       "      <td>26.057953</td>\n",
       "      <td>656.660413</td>\n",
       "      <td>638.419846</td>\n",
       "      <td>191.091655</td>\n",
       "      <td>1511.361267</td>\n",
       "      <td>425.613230</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334094</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>517.950212</td>\n",
       "      <td>517.950212</td>\n",
       "      <td>712.181542</td>\n",
       "      <td>258.975106</td>\n",
       "      <td>485.578324</td>\n",
       "      <td>453.206436</td>\n",
       "      <td>4532.064355</td>\n",
       "      <td>625.856506</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334095</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>5035.488203</td>\n",
       "      <td>1258.872051</td>\n",
       "      <td>419.624017</td>\n",
       "      <td>629.436025</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1258.872051</td>\n",
       "      <td>1258.872051</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334096</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1246.591660</td>\n",
       "      <td>415.530553</td>\n",
       "      <td>207.765277</td>\n",
       "      <td>207.765277</td>\n",
       "      <td>831.061107</td>\n",
       "      <td>1246.591660</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334097</th>\n",
       "      <td>17.645598</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>341.148233</td>\n",
       "      <td>11.763732</td>\n",
       "      <td>410.162128</td>\n",
       "      <td>105.873589</td>\n",
       "      <td>26.468397</td>\n",
       "      <td>164.692250</td>\n",
       "      <td>45.094307</td>\n",
       "      <td>94.109857</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334098</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>218.875295</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1005.002395</td>\n",
       "      <td>246.234707</td>\n",
       "      <td>34.655255</td>\n",
       "      <td>601.907060</td>\n",
       "      <td>149.564785</td>\n",
       "      <td>21.887529</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334099</th>\n",
       "      <td>60.903503</td>\n",
       "      <td>2070.719118</td>\n",
       "      <td>609.035035</td>\n",
       "      <td>213.162262</td>\n",
       "      <td>268.990474</td>\n",
       "      <td>654.712662</td>\n",
       "      <td>416.173940</td>\n",
       "      <td>3014.723422</td>\n",
       "      <td>507.529196</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR8334100</th>\n",
       "      <td>113.746417</td>\n",
       "      <td>1592.449838</td>\n",
       "      <td>909.971336</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>796.224919</td>\n",
       "      <td>682.478502</td>\n",
       "      <td>587.689821</td>\n",
       "      <td>2729.914008</td>\n",
       "      <td>454.985668</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14 rows × 1048 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            2,3-diacetamido-2,3-dideoxy-d-mannuronic acid  \\\n",
       "SRR8334087                                      49.467681   \n",
       "SRR8334088                                       0.000000   \n",
       "SRR8334089                                      77.604605   \n",
       "SRR8334090                                      27.080198   \n",
       "SRR8334091                                       0.000000   \n",
       "SRR8334092                                      21.559847   \n",
       "SRR8334093                                     364.811340   \n",
       "SRR8334094                                       0.000000   \n",
       "SRR8334095                                       0.000000   \n",
       "SRR8334096                                       0.000000   \n",
       "SRR8334097                                      17.645598   \n",
       "SRR8334098                                       0.000000   \n",
       "SRR8334099                                      60.903503   \n",
       "SRR8334100                                     113.746417   \n",
       "\n",
       "            2-O-alpha-mannosyl-D-glycerate utilization  \\\n",
       "SRR8334087                                 1879.771864   \n",
       "SRR8334088                                 1556.179701   \n",
       "SRR8334089                                 1034.728061   \n",
       "SRR8334090                                 1299.849486   \n",
       "SRR8334091                                 5327.887049   \n",
       "SRR8334092                                 2845.899764   \n",
       "SRR8334093                                    0.000000   \n",
       "SRR8334094                                  517.950212   \n",
       "SRR8334095                                    0.000000   \n",
       "SRR8334096                                    0.000000   \n",
       "SRR8334097                                    0.000000   \n",
       "SRR8334098                                    0.000000   \n",
       "SRR8334099                                 2070.719118   \n",
       "SRR8334100                                 1592.449838   \n",
       "\n",
       "            2-aminophenol Metabolism  \\\n",
       "SRR8334087                445.209126   \n",
       "SRR8334088                552.192797   \n",
       "SRR8334089                724.309642   \n",
       "SRR8334090                649.924743   \n",
       "SRR8334091               5327.887049   \n",
       "SRR8334092                344.957547   \n",
       "SRR8334093                625.390869   \n",
       "SRR8334094                517.950212   \n",
       "SRR8334095               5035.488203   \n",
       "SRR8334096                  0.000000   \n",
       "SRR8334097                341.148233   \n",
       "SRR8334098                218.875295   \n",
       "SRR8334099                609.035035   \n",
       "SRR8334100                909.971336   \n",
       "\n",
       "            2-ketoacid oxidoreductases disambiguation  \\\n",
       "SRR8334087                                 676.058302   \n",
       "SRR8334088                                 363.945253   \n",
       "SRR8334089                                 120.718274   \n",
       "SRR8334090                                  45.133663   \n",
       "SRR8334091                                   0.000000   \n",
       "SRR8334092                                 172.478774   \n",
       "SRR8334093                                  26.057953   \n",
       "SRR8334094                                 712.181542   \n",
       "SRR8334095                                1258.872051   \n",
       "SRR8334096                                1246.591660   \n",
       "SRR8334097                                  11.763732   \n",
       "SRR8334098                                   0.000000   \n",
       "SRR8334099                                 213.162262   \n",
       "SRR8334100                                   0.000000   \n",
       "\n",
       "            2-oxoglutarate dehydrogenase   2-phosphoglycolate salvage  \\\n",
       "SRR8334087                     123.669202                  321.539924   \n",
       "SRR8334088                     192.430823                  363.945253   \n",
       "SRR8334089                     564.789066                  659.639139   \n",
       "SRR8334090                     344.821183                  352.042569   \n",
       "SRR8334091                       0.000000                    0.000000   \n",
       "SRR8334092                     141.576327                  711.474941   \n",
       "SRR8334093                     656.660413                  638.419846   \n",
       "SRR8334094                     258.975106                  485.578324   \n",
       "SRR8334095                     419.624017                  629.436025   \n",
       "SRR8334096                     415.530553                  207.765277   \n",
       "SRR8334097                     410.162128                  105.873589   \n",
       "SRR8334098                    1005.002395                  246.234707   \n",
       "SRR8334099                     268.990474                  654.712662   \n",
       "SRR8334100                     796.224919                  682.478502   \n",
       "\n",
       "            3-amino-5-hydroxybenzoic Acid Synthesis  \\\n",
       "SRR8334087                               354.518378   \n",
       "SRR8334088                               332.570662   \n",
       "SRR8334089                               482.873095   \n",
       "SRR8334090                               333.989104   \n",
       "SRR8334091                                 0.000000   \n",
       "SRR8334092                               463.536704   \n",
       "SRR8334093                               191.091655   \n",
       "SRR8334094                               453.206436   \n",
       "SRR8334095                                 0.000000   \n",
       "SRR8334096                               207.765277   \n",
       "SRR8334097                                26.468397   \n",
       "SRR8334098                                34.655255   \n",
       "SRR8334099                               416.173940   \n",
       "SRR8334100                               587.689821   \n",
       "\n",
       "            5-methylaminomethyl-2-thiouridine  \\\n",
       "SRR8334087                        4996.235745   \n",
       "SRR8334088                        3714.751544   \n",
       "SRR8334089                        2819.633965   \n",
       "SRR8334090                        3709.987074   \n",
       "SRR8334091                           0.000000   \n",
       "SRR8334092                        4311.969339   \n",
       "SRR8334093                        1511.361267   \n",
       "SRR8334094                        4532.064355   \n",
       "SRR8334095                        1258.872051   \n",
       "SRR8334096                         831.061107   \n",
       "SRR8334097                         164.692250   \n",
       "SRR8334098                         601.907060   \n",
       "SRR8334099                        3014.723422   \n",
       "SRR8334100                        2729.914008   \n",
       "\n",
       "            A Hypothetical Protein Related to Proline Metabolism  \\\n",
       "SRR8334087                                         700.792142      \n",
       "SRR8334088                                         811.556080      \n",
       "SRR8334089                                         379.400289      \n",
       "SRR8334090                                         704.085138      \n",
       "SRR8334091                                        1331.971762      \n",
       "SRR8334092                                         596.489092      \n",
       "SRR8334093                                         425.613230      \n",
       "SRR8334094                                         625.856506      \n",
       "SRR8334095                                        1258.872051      \n",
       "SRR8334096                                        1246.591660      \n",
       "SRR8334097                                          45.094307      \n",
       "SRR8334098                                         149.564785      \n",
       "SRR8334099                                         507.529196      \n",
       "SRR8334100                                         454.985668      \n",
       "\n",
       "            A new toxin - antitoxin system  ...  Thermodesulfobiaceae  \\\n",
       "SRR8334087                        0.000000  ...                     0   \n",
       "SRR8334088                        0.000000  ...                     0   \n",
       "SRR8334089                       51.736403  ...                     0   \n",
       "SRR8334090                      162.481186  ...                     0   \n",
       "SRR8334091                        0.000000  ...                     0   \n",
       "SRR8334092                        0.000000  ...                     0   \n",
       "SRR8334093                        0.000000  ...                     0   \n",
       "SRR8334094                        0.000000  ...                     0   \n",
       "SRR8334095                        0.000000  ...                     0   \n",
       "SRR8334096                        0.000000  ...                     0   \n",
       "SRR8334097                       94.109857  ...                     0   \n",
       "SRR8334098                       21.887529  ...                     0   \n",
       "SRR8334099                        0.000000  ...                     0   \n",
       "SRR8334100                        0.000000  ...                     0   \n",
       "\n",
       "            Petrotogaceae  Thermotogaceae     Methylacidiphilaceae     \\\n",
       "SRR8334087              0               0  0                     0  0   \n",
       "SRR8334088              0               0  0                     0  0   \n",
       "SRR8334089              0               0  0                     0  0   \n",
       "SRR8334090              0               0  0                     0  0   \n",
       "SRR8334091              0               0  0                     0  0   \n",
       "SRR8334092              0               0  0                     0  0   \n",
       "SRR8334093              0               0  0                     0  0   \n",
       "SRR8334094              0               0  0                     0  0   \n",
       "SRR8334095              0               0  0                     0  0   \n",
       "SRR8334096              0               0  0                     0  0   \n",
       "SRR8334097              0               0  0                     0  0   \n",
       "SRR8334098              0               0  0                     0  0   \n",
       "SRR8334099              0               0  0                     0  0   \n",
       "SRR8334100              0               0  0                     0  0   \n",
       "\n",
       "            Akkermansiaceae  Verrucomicrobiaceae        \n",
       "SRR8334087                0                    0  0  0  \n",
       "SRR8334088                0                    0  0  0  \n",
       "SRR8334089                0                    0  0  0  \n",
       "SRR8334090                0                    0  0  0  \n",
       "SRR8334091                0                    0  0  0  \n",
       "SRR8334092                0                    0  0  0  \n",
       "SRR8334093                0                    0  0  0  \n",
       "SRR8334094                0                    0  0  0  \n",
       "SRR8334095                0                    0  0  0  \n",
       "SRR8334096                0                    0  0  0  \n",
       "SRR8334097                0                    0  0  0  \n",
       "SRR8334098                0                    0  0  0  \n",
       "SRR8334099                0                    0  0  0  \n",
       "SRR8334100                0                    0  0  0  \n",
       "\n",
       "[14 rows x 1048 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_id = worldwide_samples[worldwide_sample_number]\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"Analysing: {sample_id}\")\n",
    "os.makedirs(os.path.join(\"img\", \"worldwide\", sample_id, \"img\"), exist_ok=True)\n",
    "print(f\"Sample is {sample_id}\", file=sys.stderr)\n",
    "\n",
    "wwdf, wwmd = cf_analysis_lib.read_worldwide_data(sample_id, sslevel=sslevel, ss_normalisation=ss_normalisation,\n",
    "                        taxonomy=taxa, all_taxa=all_taxa, raw_taxa=False, drop_amplicon=True,\n",
    "                        drop_suspected_amplicon=True, verbose=True)\n",
    "\n",
    "# now we add any columns from our_df that are not in wwdf. This make subsequent computations easier (and correct!)\n",
    "missing = [c for c in our_df.columns if c not in wwdf.columns]\n",
    "missing_df = pd.DataFrame(0, index=wwdf.index, columns=missing)\n",
    "wwdf = pd.concat([wwdf, missing_df], axis=1).copy()\n",
    "\n",
    "print(f\"We added {len(missing)} columns to ww df that were missing\", file=sys.stderr)\n",
    "wwdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "21d2087b-981a-48eb-8dce-bb2d5795fab7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"background-color:#d4edda;border:1px solid #155724;color:#155724;padding:10px 12px;border-radius:8px;font-family:system-ui,-apple-system,Segoe UI,Roboto,Helvetica,Arial,sans-serif;line-height:1.4;max-width:none;\"><div style=\"font-weight:600; margin-bottom:4px;\">✅ Analysis Report</div><div>- The entire data set is 14 runs, and 906,511,907 bp (including amplicon sequences)</div></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# print run counts\n",
    "run_counts = {\"PRJEB14440\": \"- The entire data set is 5 runs, and 1,330,576,074 bp\", \"PRJEB20836\": \"- The entire data set is 1396 runs, and 1,007,979,034,098 bp\", \"PRJEB32062\": \"- The entire data set is 27 runs, and 113,773,728,053 bp\", \"PRJEB51171\": \"- The entire data set is 64 runs, and 29,874,922,519 bp\", \"PRJEB54014\": \"- The entire data set is 80 runs, and 447,488,092,200 bp\", \"PRJNA1055940\": \"- The entire data set is 61 runs, and 386,942,064,540 bp\", \"PRJNA1081394\": \"- The entire data set is 549 runs, and 510,156,291,748 bp\", \"PRJNA1091195\": \"- The entire data set is 44 runs, and 146,735,286,366 bp\", \"PRJNA1101448\": \"- The entire data set is 323 runs, and 89,493,823,284 bp\", \"PRJNA1126024\": \"- The entire data set is 2 runs, and 1,071,033,642 bp\", \"PRJNA316056\": \"- The entire data set is 12 runs, and 36,012,000,000 bp\", \"PRJNA316588\": \"- The entire data set is 18 runs, and 121,257,187,412 bp\", \"PRJNA510441\": \"- The entire data set is 14 runs, and 906,511,907 bp\", \"PRJNA516442\": \"- The entire data set is 93 runs, and 34,371,719,753 bp\", \"PRJNA516870\": \"- The entire data set is 79 runs, and 39,170,766,613 bp\", \"PRJNA615628\": \"- The entire data set is 71 runs, and 95,889,827,074 bp\", \"PRJNA644285\": \"- The entire data set is 12 runs, and 29,679,938,510 bp\", \"PRJNA71831\": \"- The entire data set is 38 runs, and 1,179,409,307 bp\", \"PRJNA825831\": \"- The entire data set is 117 runs, and 58,764,823,681 bp\", \"PRJNA839435\": \"- The entire data set is 12 runs, and 8,336,976,960 bp\", \"PRJNA846291\": \"- The entire data set is 98 runs, and 87,154,850,753 bp\", \"PRJNA931830\": \"- The entire data set is 260 runs, and 740,690,138,100 bp\"}\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"{run_counts[sample_id]} (including amplicon sequences)\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "16ef2ac4-9111-4665-b53d-795aca152b3f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"background-color:#d4edda;border:1px solid #155724;color:#155724;padding:10px 12px;border-radius:8px;font-family:system-ui,-apple-system,Segoe UI,Roboto,Helvetica,Arial,sans-serif;line-height:1.4;max-width:none;\"><div style=\"font-weight:600; margin-bottom:4px;\">✅ Analysis Report</div><div>- We analysed 14 metagenomic sequence runs.</div></div>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"- We analysed {len(wwdf.index)} metagenomic sequence runs.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13c7718c-c70f-4b47-8285-99d0b408454b",
   "metadata": {},
   "source": [
    "## Create the PCA\n",
    "\n",
    "Now we need to create the PCA df using our original data, and use those axis to fit our new data. We need the cluster assignments and the raw data!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4b42b427-e670-48e0-a775-3d51797be983",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['Haloferacaceae'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 9\u001b[0m\n\u001b[1;32m      7\u001b[0m ww_pc_data \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m cluster_id, group \u001b[38;5;129;01min\u001b[39;00m grouped:\n\u001b[0;32m----> 9\u001b[0m     df_clust \u001b[38;5;241m=\u001b[39m \u001b[43mour_df\u001b[49m\u001b[43m[\u001b[49m\u001b[43mcluster_assignments\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mcluster_assignments\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mCluster\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m==\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mcluster_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mFeature\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m     10\u001b[0m     ww_clust \u001b[38;5;241m=\u001b[39m wwdf[cluster_assignments\u001b[38;5;241m.\u001b[39mloc[cluster_assignments[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCluster\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m cluster_id, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFeature\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n\u001b[1;32m     11\u001b[0m     pca \u001b[38;5;241m=\u001b[39m PCA(n_components\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/core/frame.py:3902\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3900\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m is_iterator(key):\n\u001b[1;32m   3901\u001b[0m         key \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(key)\n\u001b[0;32m-> 3902\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcolumns\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m[\u001b[38;5;241m1\u001b[39m]\n\u001b[1;32m   3904\u001b[0m \u001b[38;5;66;03m# take() does not accept boolean indexers\u001b[39;00m\n\u001b[1;32m   3905\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(indexer, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mbool\u001b[39m:\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/core/indexes/base.py:6114\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6111\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6112\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6114\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6116\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6117\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6118\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m~/.local/lib/python3.9/site-packages/pandas/core/indexes/base.py:6178\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6175\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6177\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 6178\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['Haloferacaceae'] not in index\""
     ]
    }
   ],
   "source": [
    "grouped = cluster_assignments.groupby(\"Cluster\")\n",
    "\n",
    "# pc_df = pd.DataFrame(index=our_df.index, columns=[f\"Cluster {x+1}\" for x in range(len(cluster_assignments.groupby(\"Cluster\").size()))])\n",
    "# ww_pc_df = pd.DataFrame(index=wwdf.index, columns=[f\"Cluster {x+1}\" for x in range(len(cluster_assignments.groupby(\"Cluster\").size()))])\n",
    "fit_pca = {}\n",
    "pc_df_data = {}\n",
    "ww_pc_data = {}\n",
    "for cluster_id, group in grouped:\n",
    "    df_clust = our_df[cluster_assignments.loc[cluster_assignments[\"Cluster\"] == cluster_id, \"Feature\"]]\n",
    "    ww_clust = wwdf[cluster_assignments.loc[cluster_assignments[\"Cluster\"] == cluster_id, \"Feature\"]]\n",
    "    pca = PCA(n_components=1)\n",
    "    pca.fit(df_clust)\n",
    "    fit_pca[cluster_id] = pca\n",
    "    pc_df_data[f\"Cluster {cluster_id}\"] = pca.transform(df_clust)[:,0]\n",
    "    ww_pc_data[f\"Cluster {cluster_id}\"] = pca.transform(ww_clust)[:,0]\n",
    "\n",
    "pc_df = pd.DataFrame(pc_df_data, index=our_df.index)\n",
    "ww_pc_df = pd.DataFrame(ww_pc_data, index=wwdf.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94c384fc-71bb-4ff9-b35a-3208483903e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ww_pc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f8ffddb3-464b-4a88-a2b3-4944d5c13b2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Our data: min: {min(pc_df['Cluster 56']):.2f} max: {max(pc_df['Cluster 56']):.2f}\")\n",
    "print(f\"Sample {sample_id}: min: {min(ww_pc_df['Cluster 56']):.2f} max: {max(ww_pc_df['Cluster 56']):.2f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7bf54e8-f6a0-489d-92b5-1345fba1aff4",
   "metadata": {},
   "source": [
    "# Scale the data frames. \n",
    "\n",
    "We use a robust scaler since all the PCAs have different units. This normalises all the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b714d6c1-e92a-4506-88b6-1fc720226fc3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# scaler = StandardScaler() # - not good because too many values at 0\n",
    "scaler = RobustScaler() \n",
    "# scaler = MinMaxScaler() \n",
    "\n",
    "# train the scaler on pc_df and then apply the _same scalar_ to ww_pc_df\n",
    "\n",
    "trained_scaler = scaler.fit(pc_df)\n",
    "\n",
    "pc_scaled = pd.DataFrame(\n",
    "    trained_scaler.transform(pc_df),\n",
    "    index=pc_df.index,\n",
    "    columns=pc_df.columns\n",
    ")\n",
    "\n",
    "ww_pc_scaled = pd.DataFrame(\n",
    "    trained_scaler.transform(ww_pc_df),\n",
    "    index=ww_pc_df.index,\n",
    "    columns=ww_pc_df.columns\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d01cdcf-c65e-45d7-9dba-1e0124c75fff",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(f\"Our data: min: {min(pc_scaled['Cluster 56']):.2f} max: {max(pc_scaled['Cluster 56']):.2f}\")\n",
    "print(f\"Sample {sample_id}: min: {min(ww_pc_scaled['Cluster 56']):.2f} max: {max(ww_pc_scaled['Cluster 56']):.2f}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2740ba4f-060c-4871-b2ac-ecb04d7e73e9",
   "metadata": {},
   "source": [
    "# Build the GBRF with our data first\n",
    "\n",
    "This is from our separate model analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e7e6c5c-200e-402a-94e2-af6c3c13e00e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#intcol = 'Pseudomonas Culture' # note these two columns have the same information, but 'CS_Pseudomonas aeruginosa' is a category, while 'Pseudomonas culture' is a float!\n",
    "intcol = 'CS_Pseudomonas aeruginosa'\n",
    "\n",
    "intcol_title = replace_index.sub('', intcol).replace('_', ' ')\n",
    "merged_df = pc_scaled.join(our_metadata[[intcol]]).dropna(subset=intcol)\n",
    "\n",
    "categorical_data, custom_labels = cf_analysis_lib.create_custom_labels(our_metadata, intcol, merged_df)\n",
    "\n",
    "X = merged_df.drop(intcol, axis=1)\n",
    "y = merged_df[intcol]\n",
    "\n",
    "if categorical_data:\n",
    "  model, mse, feature_importances_sorted = cf_analysis_lib.gb_classifier_model(X, y, n_estimators=1000, n_iter_no_change=20)\n",
    "  met = 'classifier'\n",
    "else:\n",
    "  model, mse, feature_importances_sorted = cf_analysis_lib.gb_regressor_model(X, y)\n",
    "  met = 'regressor'\n",
    "\n",
    "print(f\"We used {model.n_estimators_} estimators for the random forest {met}\", file=sys.stderr)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28d98cc2-92c4-42b8-84a1-d948e1761bdb",
   "metadata": {},
   "source": [
    "## Apply our model to the new data\n",
    "\n",
    "We use the model from the GBRF to predict the outcomes based on the clustered data.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e05f5912-bb70-430f-b350-617493d196db",
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = model.predict(ww_pc_scaled)\n",
    "perc = predictions.sum()/len(predictions) * 100\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"- We predicted {predictions.sum()} samples out of {len(predictions)} ({perc:0.1f}%) have _Pseudomonas aeruginosa_\")\n",
    "predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b395925b-60f0-43b6-a28f-237c421108ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# how confident are we in our predictions?\n",
    "def confidence_label(prob, threshold_low=0.6, threshold_high=0.8):\n",
    "    if prob < threshold_low:\n",
    "        return \"Low\"\n",
    "    elif prob < threshold_high:\n",
    "        return \"Medium\"\n",
    "    else:\n",
    "        return \"High\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79b46722-361e-4e5a-9491-01330ed558f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "probs = model.predict_proba(ww_pc_scaled)\n",
    "print(probs[:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4319fb9b-fbb6-4619-ae5f-875e30ce9d48",
   "metadata": {},
   "outputs": [],
   "source": [
    "pred_class = np.argmax(probs, axis=1)\n",
    "pred_conf = probs[np.arange(len(probs)), pred_class]\n",
    "label_map = {0: \"Negative\", 1: \"Positive\"}\n",
    "\n",
    "# Confidence labels\n",
    "confidence = [confidence_label(p) for p in pred_conf]\n",
    "\n",
    "output = [\"Sample | Pseudomonas Prediction | Confidence | Certainty\\n --- | --- | --- | ---\"]\n",
    "\n",
    "for i, (cls, conf, label) in enumerate(zip(pred_class, pred_conf, confidence)):\n",
    "    output.append(f\"{ww_pc_scaled.index[i]} | {label_map[cls]} | {conf:.2f} | {label}\")\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=\"\\n\".join(output))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d855ab8-a3f9-4423-a6d5-8e03d9b83dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "interesting_cluster = 56\n",
    "fig, axes = plt.subplots(figsize=(24, 12), nrows=1, ncols=2, sharex=False, sharey=True)\n",
    "\n",
    "df_clust = our_df[cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"]]\n",
    "df_clust_corr = df_clust.corr()\n",
    "g = sns.heatmap(df_clust_corr, ax=axes[0], annot=False, cmap='coolwarm')\n",
    "g.set_title(f'Cluster {interesting_cluster} Correlation Matrix for our data')\n",
    "\n",
    "valid_features = [f for f in cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"] if f in wwdf.columns]\n",
    "df_clust = wwdf[valid_features]\n",
    "df_clust_corr = df_clust.corr()\n",
    "\n",
    "sns.heatmap(df_clust_corr, ax=axes[1], annot=False, cmap='coolwarm')\n",
    "plt.title(f'Cluster {interesting_cluster} Correlation Matrix for sample {sample_id}')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "088153df-ea0f-4f0c-8172-ad30aed3e1bb",
   "metadata": {},
   "source": [
    "# Plot a t-SNE of our data and the worldwide data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e9f3db3-1fdb-4fd9-b3a3-a90a91f8567a",
   "metadata": {},
   "outputs": [],
   "source": [
    "X_combined = np.vstack([pc_df, ww_pc_df])\n",
    "sample_labels = ([\"Adelaide\"] * len(our_df)) + ([sample_id] * len(wwdf))\n",
    "group_labels = list(our_metadata[intcol]) + list(predictions)\n",
    "sample_ids = list(pc_df.index) + list(ww_pc_df.index)\n",
    "\n",
    "tsne = TSNE(n_components=2, perplexity=30, random_state=42)\n",
    "tsne_result = tsne.fit_transform(X_combined)\n",
    "\n",
    "df_tsne = pd.DataFrame(tsne_result, columns=[\"TSNE1\", \"TSNE2\"])\n",
    "df_tsne[\"SampleType\"] = sample_labels  # Or use group_labels\n",
    "df_tsne[\"Pseudomonas Predictions\"] = group_labels  # Or use group_labels\n",
    "df_tsne['Pseudomonas Predictions'] = df_tsne['Pseudomonas Predictions'].map({0: 'Negative', 1: 'Positive'})\n",
    "df_tsne['Sample IDs'] = sample_ids\n",
    "\n",
    "# Basic comparison: New vs Original\n",
    "g = sns.scatterplot(data=df_tsne, x=\"TSNE1\", y=\"TSNE2\", palette='coolwarm', style=\"SampleType\", hue=\"Pseudomonas Predictions\")\n",
    "g.set_xlabel(\"t-SNE1\")\n",
    "g.set_ylabel(\"t-SNE2\")\n",
    "#plt.title(f\"t-SNE of Adelaide and {sample_id} Samples\")\n",
    "plt.title(\"\")\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1, 0.5))\n",
    "\n",
    "imgout = os.path.join(\"img\", \"worldwide\", sample_id, \"img\", f\"{sample_id}_Pseudomonas_tSNE.png\")\n",
    "plt.savefig(imgout)\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"## t-SNE\\n![Comparison of Adelaide and {sample_id} samples by t-SNE](img/{sample_id}_Pseudomonas_tSNE.png 'Fig. t-SNE of all the analysed sequence data coloured by whether Pseudomonas is predicted')\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7938039e-8209-4b52-a23d-3f9e7405bf34",
   "metadata": {},
   "source": [
    "# Create a PCA of all the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8733ada8-c8f9-467c-a8bf-0dd6b899a00d",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_model = PCA(n_components=2)\n",
    "pca_combined_result = pca_model.fit_transform(X_combined)\n",
    "\n",
    "# Put into DataFrame for plotting\n",
    "pca_combined_df = pd.DataFrame(pca_combined_result, columns=[\"PC1\", \"PC2\"])\n",
    "pca_combined_df[\"Source\"] = ([\"Adelaide\"] * len(our_df)) + ([sample_id] * len(wwdf))\n",
    "pca_combined_df[\"Predictions\"] = list(our_metadata[intcol]) + list(predictions)\n",
    "pca_combined_df[\"Predictions\"] = pca_combined_df[\"Predictions\"].map({0: 'Negative', 1: 'Positive'})\n",
    "\n",
    "pca_combined_df['IDS'] = list(our_metadata.index) + list(wwdf.index)\n",
    "\n",
    "custom_labels = {0.0: 'No', 1.0: 'Yes'}\n",
    "\n",
    "# Get the loadings (contributions of each original variable to the PCs)\n",
    "loadings = pd.DataFrame(pca_model.components_.T, \n",
    "                        index=pc_df.columns, \n",
    "                        columns=[\"PC1\", \"PC2\"])\n",
    "\n",
    "# Scale arrows for better visibility\n",
    "arrow_scale = 450000\n",
    "\n",
    "plt.figure(figsize=(8, 6))\n",
    "ax = sns.scatterplot(data=pca_combined_df, x=\"PC1\", palette='coolwarm',\n",
    "                y=\"PC2\", style=\"Source\", hue='Predictions', alpha=0.8, s=60)\n",
    "plt.title(f\"PCA Comparison of Adelaide and {sample_id} Samples\")\n",
    "plt.xlabel(f\"PC1 ({pca_model.explained_variance_ratio_[0]:.2%}%)\")\n",
    "plt.ylabel(f\"PC2 ({pca_model.explained_variance_ratio_[1]:.2%}%)\")\n",
    "# Plot top N most important loadings (e.g., based on magnitude of PC1^2 + PC2^2)\n",
    "top_features = loadings.pow(2).sum(axis=1).sort_values(ascending=False).head(10).index\n",
    "\n",
    "for feature in top_features:\n",
    "    x = loadings.loc[feature, \"PC1\"]\n",
    "    y = loadings.loc[feature, \"PC2\"]\n",
    "    plt.arrow(0, 0, x * arrow_scale, y * arrow_scale,\n",
    "              color='black', alpha=0.7, head_width=0.02)\n",
    "    plt.text(x * arrow_scale * 1.1, y * arrow_scale * 1.1, feature,\n",
    "             color='black', ha='center', va='center', fontsize=10)\n",
    "\n",
    "plt.legend(loc='center left', ncol=2, bbox_to_anchor=(0.3, -0.3))\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044d1328-d949-4f5f-b958-43fdf1690b83",
   "metadata": {},
   "source": [
    "# Plot a PCA of just the \"Pseudomonas\" Cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38174ff5-7d29-4b39-99b4-0cf382f47487",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=2)\n",
    "\n",
    "intcol = 'Pseudomonas Culture'\n",
    "valid_features = [f for f in cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"] if f in wwdf.columns]\n",
    "df_clust = wwdf[valid_features]\n",
    "\n",
    "pca_result = pca.fit_transform(df_clust)\n",
    "pca_df = pd.DataFrame(data=pca_result, index=wwdf.index, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Get loadings\n",
    "loadings = pca.components_.T * np.sqrt(pca.explained_variance_)\n",
    "loadings_df = pd.DataFrame(loadings, index=df_clust.columns, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Create a DataFrame for top loadings\n",
    "top_loadings_df = loadings_df.loc[loadings_df['PC1'].abs().sort_values(ascending=False).index]\n",
    "top_loadings_df.head()\n",
    "\n",
    "explained_variance = pca.explained_variance_ratio_ * 100\n",
    "pc1_variance = explained_variance[0]\n",
    "pc2_variance = explained_variance[1]\n",
    "\n",
    "\n",
    "# don't forget to change the legend\n",
    "intcol_neg = 0\n",
    "colours = np.where(predictions == intcol_neg, 'blue', 'red')\n",
    "\n",
    "# Plot the PCA results\n",
    "plt.figure(figsize=(10, 8))\n",
    "plt.scatter(pca_df['PC1'], pca_df['PC2'], c=colours, alpha=0.2)\n",
    "#plt.title(f\"PCA of the 'Pseudomonas' cluster (#56) in {sample_id}\")\n",
    "plt.title(\"\")\n",
    "plt.xlabel(f'Principal Component 1 ({pc1_variance:.3f}%)')\n",
    "plt.ylabel(f'Principal Component 2 ({pc2_variance:.3f}%)')\n",
    "\n",
    "# add the loadings ... we only plot maxloadings here\n",
    "maxloadings = 5\n",
    "if len(loadings) < maxloadings:\n",
    "    maxloadings = len(loadings)\n",
    "\n",
    "plotscaler = 2\n",
    "texts = []\n",
    "colour_cycle = cycle(mcolors.TABLEAU_COLORS)\n",
    "found_pseudomonas = False\n",
    "for i in range(maxloadings):\n",
    "    c = next(colour_cycle)\n",
    "    xpos = top_loadings_df.iloc[i, 0]*plotscaler\n",
    "    ypos = top_loadings_df.iloc[i, 1]*plotscaler\n",
    "    plt.arrow(0, 0, xpos, ypos, \n",
    "              color=c, alpha=0.5, width=0.05)\n",
    "    texts.append(plt.text(xpos, ypos, top_loadings_df.index[i], color=c))\n",
    "\n",
    "adjust_text(texts)\n",
    "\n",
    "\n",
    "# Add a legend\n",
    "blue_patch = plt.Line2D([0], [0], marker='o', color='w', label='Predicted Pseudomonas positive', \n",
    "                         markerfacecolor='red', alpha=0.2, markersize=10)\n",
    "red_patch = plt.Line2D([0], [0], marker='o', color='w', label='Predictied Pseudomonas negative', \n",
    "                        markerfacecolor='blue', alpha=0.2, markersize=10)\n",
    "\n",
    "\n",
    "plt.legend(handles=[blue_patch, red_patch])\n",
    "\n",
    "imgout = os.path.join(\"img\", \"worldwide\", sample_id, \"img\", f\"{sample_id}_Pseudomonas_PCA.png\")\n",
    "plt.savefig(imgout)\n",
    "cf_analysis_lib.show_green(title=\"Analysis Report\", message=f\"## PCA\\n![This cluster of features are most strongly associated with the presence of Pseudomonas](img/{sample_id}_Pseudomonas_PCA.png 'Fig. PCA of the cluster of features most strongly associated with Pseudomonas colonization in {sample_id}')\")\n",
    "# Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e727fa0e-5e54-403f-9aeb-8295908e32e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "intcol = 'Pseudomonas Culture'\n",
    "valid_features = [f for f in cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"] if f in wwdf.columns]\n",
    "df_clust = pd.concat([our_df[valid_features], wwdf[valid_features]])\n",
    "df_clust.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c2a7f6e-c91a-4841-87af-0f84b6e866bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=2)\n",
    "\n",
    "intcol = 'Pseudomonas Culture'\n",
    "valid_features = [f for f in cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"] if f in wwdf.columns]\n",
    "df_clust = pd.concat([our_df[valid_features], wwdf[valid_features]])\n",
    "\n",
    "\n",
    "pca_result = pca.fit_transform(df_clust)\n",
    "pca_df = pd.DataFrame(data=pca_result, index=df_clust.index, columns=['PC1', 'PC2'])\n",
    "\n",
    "pca_df[\"Source\"] = ([\"Adelaide\"] * len(our_df)) + ([sample_id] * len(wwdf))\n",
    "pca_df[\"Predictions\"] = list(our_metadata[intcol]) + list(predictions)\n",
    "pca_df[\"Predictions\"] = pca_df[\"Predictions\"].map({0: 'Negative', 1: 'Positive'})\n",
    "pca_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23514715-14a3-4040-b3de-d2014c7956a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=2)\n",
    "\n",
    "interesting_cluster = 70\n",
    "valid_features = [f for f in cluster_assignments.loc[cluster_assignments[\"Cluster\"] == interesting_cluster, \"Feature\"] if f in wwdf.columns]\n",
    "df_clust = pd.concat([our_df[valid_features], wwdf[valid_features]])\n",
    "\n",
    "pca_result = pca.fit_transform(df_clust)\n",
    "pca_df = pd.DataFrame(data=pca_result, index=df_clust.index, columns=['PC1', 'PC2'])\n",
    "\n",
    "pca_df[\"Source\"] = ([\"Adelaide\"] * len(our_df)) + ([sample_id] * len(wwdf))\n",
    "pca_df[\"Pseudomonas Predictions\"] = list(our_metadata[intcol]) + list(predictions)\n",
    "pca_df[\"Pseudomonas Predictions\"] = pca_df[\"Pseudomonas Predictions\"].map({0: 'Negative', 1: 'Positive'})\n",
    "\n",
    "# Get loadings\n",
    "loadings = pca.components_.T * np.sqrt(pca.explained_variance_)\n",
    "loadings_df = pd.DataFrame(loadings, index=df_clust.columns, columns=['PC1', 'PC2'])\n",
    "\n",
    "# Create a DataFrame for top loadings\n",
    "top_loadings_df = loadings_df.loc[loadings_df['PC1'].abs().sort_values(ascending=False).index]\n",
    "\n",
    "explained_variance = pca.explained_variance_ratio_ \n",
    "pc1_variance = explained_variance[0]\n",
    "pc2_variance = explained_variance[1]\n",
    "\n",
    "# Scale arrows for better visibility\n",
    "arrow_scale = 5\n",
    "\n",
    "plt.figure(figsize=(12,12))\n",
    "ax = sns.scatterplot(data=pca_df, x=\"PC1\", palette='coolwarm',\n",
    "                y=\"PC2\", style=\"Source\", hue='Pseudomonas Predictions', alpha=0.8, s=60)\n",
    "plt.title(f\"PCA Comparison of Adelaide and {sample_id} Samples cluster {interesting_cluster}\")\n",
    "plt.xlabel(f\"PC1 ({pc1_variance:.2%})\")\n",
    "plt.ylabel(f\"PC2 ({pc2_variance:.2%})\")\n",
    "# Plot top N most important loadings (e.g., based on magnitude of PC1^2 + PC2^2)\n",
    "\n",
    "for feature in top_loadings_df.index[:5]:\n",
    "    x = loadings_df.loc[feature, 'PC1']\n",
    "    y = loadings_df.loc[feature, 'PC2']\n",
    "    plt.arrow(0, 0, x * arrow_scale, y * arrow_scale,\n",
    "              color='black', alpha=0.7, head_width=0.02)\n",
    "    plt.text(x * arrow_scale * 1.1, y * arrow_scale * 1.1, feature,\n",
    "             color='black', ha='center', va='center', fontsize=10)\n",
    "\n",
    "plt.legend(loc='center left', ncol=2, bbox_to_anchor=(0.3, -0.3))\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18b0fca3-f069-4765-bd1d-30f58aa0efd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "wwdf['Pseudomonadaceae']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d77d655d-d58b-41e9-bc27-c17de899135d",
   "metadata": {},
   "source": [
    "## Just compare cluster 56"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87117d73-bcd3-41f7-878a-24364a89caf9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_df = pd.DataFrame({\n",
    "    \"Value\": pd.concat([pc_scaled[\"Cluster 56\"], ww_pc_scaled[\"Cluster 56\"]], axis=0),\n",
    "    \"Dataset\": ([\"Adelaide\"] * len(pc_scaled)) + ([sample_id] * len(ww_pc_scaled))\n",
    "})\n",
    "\n",
    "\n",
    "plt.figure(figsize=(8, 6))\n",
    "\n",
    "# Choose one:\n",
    "# Boxplot + jitter\n",
    "#sns.boxplot(data=plot_df, x=\"Dataset\", y=\"Value\", whis=1.5, palette=\"pastel\")\n",
    "#sns.stripplot(data=plot_df, x=\"Dataset\", y=\"Value\", jitter=True, color=\"k\", alpha=0.6)\n",
    "\n",
    "# Or Violin plot + jitter\n",
    "sns.violinplot(data=plot_df, x=\"Dataset\", palette='BrBG', y=\"Value\", inner=None)\n",
    "sns.stripplot(data=plot_df, x=\"Dataset\", y=\"Value\", jitter=True, color=\"k\", alpha=0.6)\n",
    "\n",
    "plt.title(\"Comparison of Cluster 56 Values\")\n",
    "plt.ylabel(\"Standardized PCA Value\")\n",
    "plt.xlabel(\"\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3218f59b-69df-48bc-802a-bc867a839e1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in wwdf.columns:\n",
    "    if \"pseudom\" in c.lower():\n",
    "        print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2f493a8-4f82-4706-ae48-f80675b83b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "wwdf.loc['SRR8334087', 'Pseudomonadaceae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "494f8c9e-80f4-4579-825e-88a1bc1fac5b",
   "metadata": {},
   "outputs": [],
   "source": [
    "ww_pc_df.loc['SRR8334087', 'Cluster 56']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7853132-a2a1-4579-850f-4537c18c6ed2",
   "metadata": {},
   "source": [
    "## The problem.\n",
    "\n",
    "Sample SRR8334087 is predicted as not having _Pseudomonas_:\n",
    "\n",
    "SRR8334087 | Negative | 0.69 | Medium\n",
    "\n",
    "However, \n",
    "\n",
    "```\n",
    "ww_pc_df.loc['SRR8334087', 'Cluster 56'] = 8644.405853116295\n",
    "and\n",
    "wwdf.loc['SRR8334087', 'Pseudomonadaceae'] = 31044.084329222947\n",
    "```\n",
    "\n",
    "so both of these are strong signals!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef2029d1-ce33-4eb2-a3ce-f9302ab2a188",
   "metadata": {},
   "outputs": [],
   "source": [
    "import shap\n",
    "\n",
    "# 1. Create a SHAP explainer for your trained model\n",
    "explainer = shap.TreeExplainer(model)\n",
    "\n",
    "# 2. Pick the sample you want to explain\n",
    "run_id = \"SRR8334087\"\n",
    "sample_index = ww_pc_df.index.get_loc(run_id)\n",
    "\n",
    "# 3. Get shap values (no extra class axis)\n",
    "shap_values = explainer.shap_values(ww_pc_scaled)  # shape: (n_samples, n_features)\n",
    "\n",
    "# 4. Base value is a single scalar\n",
    "base_value = explainer.expected_value\n",
    "\n",
    "# 5. Extract the row for our sample\n",
    "sample_shap_values = shap_values[sample_index]\n",
    "sample_features = ww_pc_scaled.iloc[sample_index]\n",
    "\n",
    "# 6. Waterfall plot\n",
    "shap.waterfall_plot(\n",
    "    shap.Explanation(\n",
    "        values=sample_shap_values,\n",
    "        base_values=base_value,\n",
    "        data=sample_features,\n",
    "        feature_names=ww_pc_scaled.columns\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7258fd1c-a109-4b7e-9445-c70eec8540f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axes = plt.subplots(figsize=(24, 24), nrows=2, ncols=2)\n",
    "# SRR8334087 is wrong\n",
    "# SRR8334093 is correct\n",
    "sample_ids = ['SRR8334087', 'SRR8334093', '1128691_20171206_S', '1128691_20171218_S']\n",
    "\n",
    "explainer = shap.TreeExplainer(model)\n",
    "shap_values = [explainer.shap_values(ww_pc_scaled), explainer.shap_values(pc_scaled)]\n",
    "base_value = explainer.expected_value\n",
    "\n",
    "for i, (ax, run_id) in enumerate(zip(axes.ravel(), sample_ids)):\n",
    "    print(i)\n",
    "    plt.sca(ax)  # set current axes\n",
    "    if i < 2:\n",
    "        sample_index = ww_pc_scaled.index.get_loc(run_id)\n",
    "        shapdata = ww_pc_scaled.iloc[sample_index]\n",
    "        shapnames = ww_pc_scaled.columns\n",
    "        shapvals = shap_values[0][sample_index]\n",
    "    else:\n",
    "        sample_index = pc_scaled.index.get_loc(run_id)\n",
    "        shapdata = pc_scaled.iloc[sample_index]\n",
    "        shapnames = pc_scaled.columns\n",
    "        shapvals = shap_values[1][sample_index]\n",
    "        \n",
    "    shap.waterfall_plot(\n",
    "        shap.Explanation(\n",
    "            values=shapvals,\n",
    "            base_values=base_value,\n",
    "            data=shapdata,\n",
    "            feature_names=shapnames\n",
    "        ),\n",
    "        max_display=10,\n",
    "        show=False\n",
    "    )\n",
    "    ax.set_title(run_id)\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01b31bcf-2bbe-4f3d-b0f3-3eb2eada9341",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_assignments[cluster_assignments['Cluster'] == 70]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f05d8524-ecf3-49af-b4d6-166f1d51e709",
   "metadata": {},
   "source": [
    "# Read the raw data, not the normalised data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "942c53bf-0570-4141-92cf-aba646c7111e",
   "metadata": {},
   "outputs": [],
   "source": [
    "ssnorm = 'raw'\n",
    "rawt = True\n",
    "wwraw, wwrawmd = cf_analysis_lib.read_worldwide_data(sample_id, sslevel=sslevel, ss_normalisation=ssnorm,\n",
    "                        taxonomy=taxa, all_taxa=all_taxa, raw_taxa=rawt, drop_amplicon=True,\n",
    "                        drop_suspected_amplicon=True, verbose=True)\n",
    "\n",
    "wwraw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a940d25-4f58-46b0-a414-9e03314b3c76",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in wwraw:\n",
    "    if 'pseudo' in c.lower():\n",
    "        print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfdccba2-04f1-434b-8d65-b2d8312bf929",
   "metadata": {},
   "outputs": [],
   "source": [
    "wwraw['Pseudomonadaceae']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ff025d8-306a-43a4-a981-4efbc21ee694",
   "metadata": {},
   "outputs": [],
   "source": [
    "ssnorm = 'raw'\n",
    "rawt = True\n",
    "taxa=\"species\"\n",
    "wwraw, wwrawmd = cf_analysis_lib.read_worldwide_data(sample_id, sslevel=sslevel, ss_normalisation=ssnorm,\n",
    "                        taxonomy=taxa, all_taxa=True, raw_taxa=rawt, drop_amplicon=True,\n",
    "                        drop_suspected_amplicon=True, verbose=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74676fde-f063-49ec-98fc-21a045fa9fb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in wwraw:\n",
    "    if 'pseudo' in c.lower():\n",
    "        print(c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3d2476b-25d4-4248-a1f4-c76a03ca5c73",
   "metadata": {},
   "outputs": [],
   "source": [
    "wwraw['Pseudomonas aeruginosa']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5a783a0-9d11-448a-9489-a682543041f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BioinfieConda",
   "language": "python",
   "name": "bioinfieconda"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
