# Bioproject PRJNA615628

This sample comes from the [NCBI BioProject PRJNA615628](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA615628).

# Title:

Cystic fibrosis airway respiratory microbiome Raw sequence reads


# Description:


Cystic fibrosis airway microbiome

# Metadata:
We have included the metadata in two separate files, a [JSON format file](PRJNA615628.metadata.json.gz) that is computer readable, or a [tsv format file](PRJNA615628.metadata.tsv.gz) that you can import into excel or similar.

# Publications


Inam Z, Felton E, Burrell A, Chaney H, Sami I, Koumbourlis AC, Freishtat RJ, Zemanick ET, Crandall KA, Hahn A. 2022. Impact of antibiotics on the lung microbiome and lung function in children with cystic fibrosis 1 year after hospitalization for an initial pulmonary exacerbation. [Open Forum Infect Dis 9:ofac466](https://doi.org/10.1093/ofid/ofac466).
  


Hahn A, Burrell A, Chaney H, Sami I, Koumbourlis AC, Freishtat RJ, Zemanick ET, Louie S, Crandall KA. 2021. Importance of beta-lactam pharmacokinetics and pharmacodynamics on the recovery of microbial diversity in the airway of persons with cystic fibrosis. [J Investig Med 69:1350–1359](https://doi.org/10.1136/jim-2021-001824).
  

Felton E, Burrell A, Chaney H, Sami I, Koumbourlis AC, Freishtat RJ, Crandall KA, Hahn A. 2021. Inflammation in children with cystic fibrosis: contribution of bacterial production of long-chain fatty acids. [Pediatr Res 90:99–108](https://doi.org/10.1038/s41390-021-01419-4).
  
# Analysis:

- The entire data set is 71 runs, and 95,889,827,074 bp
- We analysed 71 metagenomic sequence runs.
- We predicted 2 samples out of 71 (2.8%) have _Pseudomonas aeruginosa_

<details>
<summary>
Our prediction outcomes
</summary>

## Prediction outcomes

Sample | Pseudomonas Prediction | Confidence | Certainty
 --- | --- | --- | ---
SRR11440759 | Negative | 0.72 | Medium
SRR11440760 | Negative | 0.83 | High
SRR11440761 | Negative | 0.73 | Medium
SRR11440762 | Negative | 0.75 | Medium
SRR11440763 | Negative | 0.85 | High
SRR11440764 | Negative | 0.78 | Medium
SRR11440765 | Negative | 0.83 | High
SRR11440766 | Negative | 0.87 | High
SRR11440767 | Negative | 0.85 | High
SRR11440768 | Negative | 0.87 | High
SRR11440769 | Negative | 0.85 | High
SRR11440770 | Negative | 0.85 | High
SRR11440771 | Negative | 0.80 | High
SRR11440772 | Negative | 0.84 | High
SRR11440773 | Negative | 0.73 | Medium
SRR11440774 | Negative | 0.86 | High
SRR11440775 | Negative | 0.86 | High
SRR11440776 | Negative | 0.72 | Medium
SRR11440777 | Negative | 0.83 | High
SRR11440778 | Negative | 0.83 | High
SRR11440779 | Negative | 0.83 | High
SRR11440780 | Positive | 0.73 | Medium
SRR11440781 | Negative | 0.70 | Medium
SRR11440782 | Negative | 0.83 | High
SRR11440783 | Negative | 0.79 | Medium
SRR11440784 | Negative | 0.67 | Medium
SRR11440785 | Negative | 0.66 | Medium
SRR11440786 | Negative | 0.83 | High
SRR11440787 | Negative | 0.82 | High
SRR11440788 | Negative | 0.80 | Medium
SRR11440789 | Negative | 0.81 | High
SRR11440790 | Negative | 0.57 | Low
SRR11440791 | Negative | 0.83 | High
SRR11440792 | Negative | 0.65 | Medium
SRR11440793 | Negative | 0.61 | Medium
SRR11440794 | Negative | 0.86 | High
SRR11440795 | Negative | 0.78 | Medium
SRR11440796 | Negative | 0.77 | Medium
SRR11440797 | Negative | 0.82 | High
SRR11440798 | Negative | 0.81 | High
SRR11440799 | Negative | 0.75 | Medium
SRR11440800 | Negative | 0.77 | Medium
SRR11440801 | Positive | 0.54 | Low
SRR11440802 | Negative | 0.80 | Medium
SRR11440803 | Negative | 0.72 | Medium
SRR11440804 | Negative | 0.86 | High
SRR11440805 | Negative | 0.82 | High
SRR11440806 | Negative | 0.61 | Medium
SRR11440807 | Negative | 0.57 | Low
SRR11440808 | Negative | 0.88 | High
SRR11440809 | Negative | 0.75 | Medium
SRR11440810 | Negative | 0.79 | Medium
SRR11440811 | Negative | 0.86 | High
SRR11440812 | Negative | 0.68 | Medium
SRR11440813 | Negative | 0.79 | Medium
SRR11440814 | Negative | 0.82 | High
SRR11440815 | Negative | 0.69 | Medium
SRR11440816 | Negative | 0.82 | High
SRR11440817 | Negative | 0.82 | High
SRR11440818 | Negative | 0.81 | High
SRR11440819 | Negative | 0.80 | High
SRR11440820 | Negative | 0.74 | Medium
SRR11440821 | Negative | 0.79 | Medium
SRR11440822 | Negative | 0.75 | Medium
SRR11440823 | Negative | 0.84 | High
SRR11440824 | Negative | 0.74 | Medium
SRR11440825 | Negative | 0.83 | High
SRR11440826 | Negative | 0.75 | Medium
SRR11440827 | Negative | 0.81 | High
SRR11440828 | Negative | 0.82 | High
SRR11440829 | Negative | 0.78 | Medium

</details>

## t-SNE
![Comparison of Adelaide and PRJNA615628 samples by t-SNE](img/PRJNA615628_Pseudomonas_tSNE.png 'Fig. t-SNE of all the analysed sequence data coloured by whether Pseudomonas is predicted')


## PCA
![This cluster of features are most strongly associated with the presence of Pseudomonas](img/PRJNA615628_Pseudomonas_PCA.png 'Fig. PCA of the cluster of features most strongly associated with Pseudomonas colonization in PRJNA615628')


